File: breakout_zig/src/ball.zig
const rl = @import("raylib.zig");
const std = @import("std");
const collision = @import("collision.zig");
const Box = @import("box.zig").Box;
const Scene = @import("scene.zig").Scene;
const Platform = @import("platform.zig").Platform;
const CratePack = @import("crate.zig").CratePack;

const Vector2 = rl.Vector2;
const Color = rl.Color;

const Collider = collision.Collider;
const Collision = collision.Collision;

pub const Ball = struct {
    const Self = @This();

    box: Box,
    velocity: Vector2,
    speed: f32,

    pub fn new(center: Vector2, size: Vector2, velocity: Vector2, speed: f32, color: Color) Self {
        return Self{
            .box = Box.new(
                center,
                size,
                color,
            ),
            .velocity = velocity,
            .speed = speed,
        };
    }

    pub fn draw(self: *const Self) void {
        self.box.draw();
    }

    pub fn collides(self: *const Self, collider: *const Collider) ?Collision {
        return self.box.collides(collider);
    }

    pub fn update(self: *Self, scene: *const Scene, platform: *const Platform, crate_pack: *CratePack, dt: f32) void {
        self.box.collider.rect.move(Vector2{ .x = self.velocity.x * self.speed * dt, .y = self.velocity.y * self.speed * dt });

        self.check_collision_const(Scene, scene);
        self.check_collision_const(Platform, platform);
        self.check_collision(CratePack, crate_pack);
    }

    fn check_collision_const(self: *Self, comptime T: type, other: *const T) void {
        const col = other.collides(&self.box.collider);
        if (col) |c| {
            self.apply_collision(c);
        }
    }

    fn check_collision(self: *Self, comptime T: type, other: *T) void {
        const col = other.collides(&self.box.collider);
        if (col) |c| {
            self.apply_collision(c);
        }
    }

    fn apply_collision(self: *Self, c: Collision) void {
        if (c.normal.x != 0.0) {
            self.velocity.x *= -1.0;
        }
        if (c.normal.y != 0.0) {
            self.velocity.y *= -1.0;
        }
    }
};

File: breakout_go/src/ball.go
package src

import (
	rl "github.com/gen2brain/raylib-go/raylib"
)

type Ball struct {
	rectangle Rect
	color     rl.Color
	velocity  rl.Vector2
	speed     float32
}

func NewBall(center rl.Vector2, width float32, height float32, color rl.Color, velocity rl.Vector2, speed float32) Ball {
	var rectangle = rect_from_center(center, width, height)
	return Ball{
		rectangle: rectangle,
		color:     color,
		velocity:  velocity,
		speed:     speed,
	}
}

func (self *Ball) Draw() {
	rl.DrawRectangleRec(self.rectangle.raylib_rect(), self.color)
}

func (self *Ball) Update(scene *SceneBorder, platform *Platform, crate_pack *CratePack, dt float32) {
	self.rectangle.X += self.velocity.X * self.speed * dt
	self.rectangle.Y += self.velocity.Y * self.speed * dt

	{
		var collision_opt = scene.collides(self)
		if collision_opt.is_some() {
			var collision = collision_opt.unwrap()
			self.handle_collision(collision)
		}
	}

	{
		var collision_opt = platform.collides(self)
		if collision_opt.is_some() {
			var collision = collision_opt.unwrap()
			self.handle_collision(collision)
		}
	}

	{
		var collision_opt = crate_pack.collides(self)
		if collision_opt.is_some() {
			var collision = collision_opt.unwrap()
			self.handle_collision(collision)
		}
	}
}

func (self *Ball) handle_collision(collision Collision) {
	if collision.normal.X != 0.0 {
		self.velocity.X *= -1.0
	}
	if collision.normal.Y != 0.0 {
		self.velocity.Y *= -1.0
	}
}

// impl Collider
func (self *Ball) rect() Option[*Rect] {
	return option_some(&self.rectangle)
}

func (self *Ball) collides(other Collider) Option[Collision] {
	return option_none[Collision]()
}
